### PR TITLE
[Endpoint] badge

### LINKED ISSUE #204: NuGet badges for multiple packages
Some projects/repos, e.g. https://github.com/scriptcs/scriptcs, produce a number of NuGet packages, e.g. https://www.nuget.org/profiles/scriptcs/.

It would be nice to be have NuGet badges which took all packages into account, certainly for download count.

Version would also work if the versions of all the packages are the same. If they differ, perhaps the badge could display the latest version, a version range, or just '4 packages'.

I think the name of each package would have to be specified in the URL, e.g. `http://img.shields.io/nuget/multi/dt/shields.svg?ScriptCs.Contracts&ScriptCs.Core&ScriptCs.Engine.Roslyn&ScriptCs.Hosting`

Specifying the packages using a filter is difficult since the only options are search, author and tags. The only one under definitive control of the publisher is author, but even that will only be narrowly useful since most projects don't create an author specifically for their package publishing. Even if they did, it would be fairly common that they would publish packages from many repos which means the badge wouldn't be useful for puttting on a repo specific readme. Still, that could also provided as an option, e.g. `http://img.shields.io/nuget/profile/dt/scriptcs.svg`.

I guess another badge could give package count, e.g. `http://img.shields.io/nuget/multi/pc/shields.svg?ScriptCs.Contracts&ScriptCs.Core&ScriptCs.Engine.Roslyn&ScriptCs.Hosting` and `http://img.shields.io/nuget/profile/pc/scriptcs.svg`.

### PR DESCRIPTION
**Pull request by paulmelnikow**
This reimplements the idea @bkdotcom came up with in #1519, and took a stab at in #1525. It’s a really powerful way to add all sorts of custom badges, particularly considering [tools like RunKit endpoints and Jupyter Kernel Gateway](https://github.com/badges/shields/issues/2259#issuecomment-444186589), not to mention all the other ways cloud functions can be deployed these days.

I gave this a name which did not have JSON in it, so as not to confuse it with the dynamic JSON badges.

There are a few things that need doing before this can be shipped, including supporting several of the options I included in the schema.

- [x] labelColor (#2740)
- [x] isError (#2693)
- [ ] ~~link (I honestly don't know much about our link support; should we have leftLink and rightLink?)~~ (Removed for this iteration)
- [x] namedLogo, logoSvg (#2796)
    - [x] test of logoSvg
- [x] logoColor, logoWidth, logoPosition (#2796)
    - [x] test of logoWidth, ~~logoPosition~~
- [x] style
- [x] cacheSeconds (#2755)
    - [x] Service test for cacheSeconds

Alternatively, they could be removed from the schema and added later on.

Curious for feedback! And thought I would open this up for comments before I got too far along.

Ref #1534 #2259. Closes #204 #358 #742 #999 #1519 #1752 #2625.

### CODE REVIEW

**Comment by calebcartwright** [comment]
I don't really have any useful feedback on this one yet. React is wayyyy too knew for me and I'm still trying to get a feel for that endpoint service. 

Will try to get back to this one hopefully soon once I've read up a bit

**Comment by paulmelnikow** [comment]
> React is wayyyy too new for me and I'm still trying to get a feel for that endpoint service.

Probably understanding the React code isn't that important here as it's just documentation. Looking at the review app would probably be sufficient.

**Comment by calebcartwright** [comment]
Okay I think I've found it now 😄 I was looking for an `endpoint` badge via the search function as well as perusing the page but couldn't find it. 

I was able to get it via direct uri at https://shields-staging-pr-2473.herokuapp.com/#/endpoint

Can I point this thing to any endpoint that returns any JSON response?

**Comment by paulmelnikow** [comment]
> Can I point this thing to any endpoint that returns any JSON response?

Yea, that's exactly the idea! 😁Or rather, a JSON response in the specified format.

This is one I created for #2259: https://runkit.com/melnikow/version-from-requirements-txt

(Oops, that formatting will need some work.)

**Comment by calebcartwright** [comment]
Gotcha. So this is conceptually similar to the dynamic badges, except that it offers the users the ability to define their own processing functionality

**Comment by calebcartwright** [comment]
Do we have any idea how we might define guidelines recommendations for when a user should use this type of badge vs. implementing their own service in the Shields codebase and submitting that as a PR?

**Comment by paulmelnikow** [comment]
> Gotcha. So this is conceptually similar to the dynamic badges, except that it offers the users the ability to define their own processing functionality

Right.

It also allows a service to publish a prerendered badge which can be proxied through shields.io to provide formatting flexibility.


> Do we have any idea how we might define guidelines recommendations for when a user should use this type of badge vs. implementing their own service in the Shields codebase and submitting that as a PR?

It's a great way to get something going quickly and to quickly iterate and remix.

Beyond that I'd say if folks have something that's working well and they think it's generally useful, they should open an issue to suggest it. I don't have clear guidelines of what's too specific or too generic to include.


We probably wouldn't merge "is such-and-such the current month or not" (#1519) or "how many days are left in the year," however I could see building something along those lines that's specific to hacktoberfest.

This is also a good candidate for things we don't want to do, like exact star count or download count (#742), or nuget aggregate download counts (#204).

**Comment by calebcartwright** [comment]
> It also allows a service to publish a prerendered badge which can be proxied through shields.io to provide formatting flexibility. 

Nice! I was thinking along the same lines of use cases. I was envisioning a scenario where someone may use this to get something going quickly, but then once they get it working well they just use it as is and don't not bother to share it back with the community 😄 

Obviously that's not something that can be prevented, but maybe just a note in docs down the line that includes the exact sentiment you articulated:

> Beyond that I'd say if folks have something that's working well and they think it's generally useful, they should open an issue to suggest it

**Comment by paulmelnikow** [comment]
> I was envisioning a scenario where someone may use this to get something going quickly, but then once they get it working well they just use it as is and not bother to share it back with the community 😄

Right, that's a possibility. Though the advantage of pulling the implementation into the shields server is that it will take load off their server endpoint, and probably will perform better and be more reliable.

**Comment by calebcartwright** [comment]
One odd thing I've noticed is that the first time I hit the endpoint page I see what I showed above in https://github.com/badges/shields/pull/2473#issuecomment-451310460, but after I reload the page I see this:

image

**Comment by paulmelnikow** [comment]
Another guideline of when you might want to use an endpoint badge: for a service which is too slow or unreliable for Shields to advertise it.

**Comment by paulmelnikow** [comment]
I'm thinking about deferring link support in the endpoint badge in order to give more thought to what Shields should really offer, and how to best design the API in accordance.

**Comment by paulmelnikow** [comment]
This is getting really close!

**Comment by paulmelnikow** [comment]
Other than the frontend formatting issue, I think this is ready to go.

I'm running into an interop issue between styled-components and Next.js SSR. I've spent a fair amount of time trying to fix it and don't think I should spend any more given #1943.

We could either block on the formatting fix, or open a bug ticket and fix the formatting in a follow-on.

The formatting doesn't look great, but it's functional, and probably people who are interested in the functionality would forgive us for a week of bad formatting. In the interest of keeping things moving, and avoiding creating conflicts in this PR as we implement #1943, I'd lean toward merging and opening a bug ticket for the fix.

File: lib/validate.js [hunk 1]
```diff
@@ -11,17 +11,21 @@ function validate(
     includeKeys = false,
     traceErrorMessage = 'Data did not match schema',
     traceSuccessMessage = 'Data after validation',
+    allowAndStripUnknownKeys = true,
```
Comment (calebcartwright) [inline lib/validate.js]: i have no clue what all this is used for, but my first thought here was whether any consumers of this function would ever need different values for these two keys (allowUnknown false and stripUnknown true)?

File: services/cache-headers.spec.js [hunk 2]
```diff
@@ -35,6 +35,11 @@ describe('Cache header functions', function() {
         serviceDefaultCacheLengthSeconds: 900,
         queryParams: { maxAge: 1000 },
       }).expect(1000)
+      given({
+        cacheHeaderConfig,
+        serviceDefaultCacheLengthSeconds: 900,
```
Comment (calebcartwright) [inline services/cache-headers.spec.js]: Would it be feasible here to update the naming conventions used here (and in cache-headers.js`) to match with the new naming conventions used in the base service (basically dropping `Length` I believe)

File: services/endpoint/endpoint.service.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+const blockedDomains = ['github.com', 'shields.io']
+
+const queryParamSchema = Joi.object({
+  url: optionalUrl.required(),
+}).required()
+
+const anySchema = Joi.any()
+
+const optionalStringWhenNamedLogoPrsent = Joi.alternatives().when('namedLogo', {
```
Comment (calebcartwright) [inline services/endpoint/endpoint.service.js]: nitpick here on typo optionalStringWhenNamedLogoPrsent -> optionalStringWhenNamedLogoPr**e**sent

File: services/endpoint/endpoint.service.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+  async handle(namedParams, queryParams) {
+    const { url } = this.constructor._validateQueryParams(
+      queryParams,
+      queryParamSchema
+    )
+
+    const { protocol, hostname } = new URL(url)
+    if (protocol !== 'https:') {
+      throw new InvalidParameter({ prettyMessage: 'please use https' })
```
Comment (calebcartwright) [inline services/endpoint/endpoint.service.js]: 👍

**Comment by calebcartwright** [comment]
I went through the changes and left feedback/questions where I could, but have to defer on the frontend content to someone with a better eye for that sort of thing. Bravo on another big add 🍻! 

> I'd lean toward merging and opening a bug ticket for the fix. 

That makes sense to me. I feel like there's been a lot of issues opened lately that will be resolved by this so I definitely see value in getting the functionality out even if there's some short term formatting items.

File: services/cache-headers.spec.js (hunk 2 above)
Comment (paulmelnikow) [reply services/cache-headers.spec.js]: Yea, I dropped `Length` which doesn't seem to add anything (cache seconds == cache length seconds). Agreed that cleaning up the rest of them is a nice idea, though I'd rather deal with it in a follow-on. `BaseService` has a property called `_cacheLength` that probably should be renamed to `cacheSeconds` and that's bound to affect a handful of files and code that's spread about a few places.

File: lib/validate.js (hunk 1 above)
Comment (paulmelnikow) [reply lib/validate.js]: It seems unsafe to `allowUnknown` and not `stripUnknown` so if we did that, I'd tend to think we weren't doing a good job of validating.

Originally I'd passed `options` directly through which would have supported any combination, though I'm not sure we would ever want to do that.

**Comment by paulmelnikow** [comment]
I'm going to add a **Beta** label to this feature, which will give us a chance to change the API if we decide we want to.

File: services/cache-headers.spec.js (hunk 2 above)
Comment (calebcartwright) [reply services/cache-headers.spec.js]: Gotcha and agreed 👍 When I first saw this I thought I was still up in the BaseService file, took me a second to remember I'd moved on to another file

File: frontend/components/endpoint-page.js [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    <Header />
+    <H3 id="static-badge">Endpoint (Beta)</H3>
+    <Snippet snippet={`${baseUrl}/badge/endpoint.svg?url=...&style=...`} />
+    <p>Endpoint response:</p>
+    <JsonExample
+      data={{
+        schemaVersion: 1,
+        label: 'hello',
+        message: 'sweet world',
+        color: 'orange',
```
Comment (calebcartwright) [inline frontend/components/endpoint-page.js]: 😆 For some reason my brain read this as `hello sweet orange`. wonder if i am hungry...

**Comment by paulmelnikow** [comment]
Do you mean you don't have much to offer on the way it's coded, or about what it's saying to the user?

**Comment by paulmelnikow** [comment]
Opened #2837 for the formatting issues. I'm gonna ship this!!!

**Comment by paulmelnikow** [comment]
And opened #2838 for further discussions.

**Pull request merged**